phoenix-core/src/main/java/org/apache/phoenix/optimize/QueryOptimizer.java

```diff
@@ -312,6 +317,16 @@ private static QueryPlan addPlan(PhoenixStatement statement, SelectStatement sel
         return null;
     }
 
+    // returns true if we can still use the index
+    // retuns false if we've been in PENDING_DISABLE too long - index should be considered disabled
+    private static boolean isUnderPendingDisableThreshold(PhoenixStatement statement, PTable indexTable) {
+        return EnvironmentEdgeManager
+                .currentTimeMillis() - indexTable.getIndexDisableTimestamp() <= statement
```

Comment:
@JamesRTaylor
> we should store the client timestamp when we first received the exception

What about cases where the client never got the exception, though?  That's actually the case that this code is meant to address - if for some reason a client is writing and then goes away before receiving the IndexWriteException, we want new/other client threads to not use the index if the index is in PENDING_DISABLE for too long.

Reply:
Good point. So instead of using EnvironmentEdgeManager.currentTimeMillis(), use indexTableRef.getCurrentTime(). That'll be the server-side timestamp when the index table was resolved during query compilation. The only caveat is that when UPDATE_CACHE_FREQUENCY is set, we wouldn't ping the server so we wouldn't have a timestamp. There are numerous issues related to this already, though, so I think this would be fine. For example, if we don't ping the server, we wouldn't know that the index is in a PENDING_DISABLE state either. We already have a JIRA on this, so we can tackle these issues down-the-road (and we're not using UPDATE_CACHE_FREQUENCY yet).

Reply:
@JamesRTaylor I'm seeing the following issue with using indexTableRef.getCurrentTime().  In my test, if set the index to PENDING_DISABLE.  I then perform a query, which uses the index as expected.  If I then advance the EnvironmentEdgeManager clock to beyond the threshold, then perform the query again, it still uses the index because indexTableRef.getCurrentTime() hasn't changed.  Is this expected?

Reply:
If you're rerunning the query, the current time should change. You should have a new TableRef for the new query (the old one won't change). The timestamp should come from the client-side cache and should be updated in MetaDataClient.updateCache() here:

    // if we aren't adding the table, we still need to update the resolved time of the table
    connection.updateResolvedTimestamp(table, resolvedTime);

The updateCache call should be made by FromCompiler.BaseColumnResolver.createTableRef() and the timestamp in the TableRef should be set based on the one in MetaDataMutationResult.

Reply:
@JamesRTaylor So found out the problem is that for index tables that are disabled, in MetaDataEndpointImpl we use the the indexDisableTimestamp for the mutationTime.  There's this comment:
"            // Freeze time for table at min non-zero value of INDEX_DISABLE_TIMESTAMP
            // This will keep the table consistent with index as the table has had one more
            // batch applied to it."
Is that still required?   I think I have to find another way to get the server time because of this.

Reply:
The timestamp freeze should only happen when QueryServices.INDEX_FAILURE_BLOCK_WRITE is true. It's false by default. There might be a bug here.